src/expressions/OrderByExpression.ts

```diff
@@ -0,0 +1,343 @@
+import atomize from './dataTypes/atomize';
+import ISequence from './dataTypes/ISequence';
+import sequenceFactory from './dataTypes/sequenceFactory';
+import Value from './dataTypes/Value';
+import DynamicContext from './DynamicContext';
+import ExecutionParameters from './ExecutionParameters';
+import Expression, { RESULT_ORDERINGS } from './Expression';
+import Specificity from './Specificity';
+import StaticContext from './StaticContext';
+import {
+	DONE_TOKEN,
+	IAsyncIterator,
+	IterationHint,
+	IterationResult,
+	ready,
+} from './util/iterators';
+import { errXUST0001 } from './xquery-update/XQueryUpdateFacilityErrors';
+
+import castToType from './dataTypes/castToType';
+import isSubtypeOf from './dataTypes/isSubtypeOf';
+import { getPrimitiveTypeName } from './dataTypes/typeHelpers';
+import FlworExpression from './FlworExpression';
+import valueCompare from './operators/compares/valueCompare';
+import PossiblyUpdatingExpression from './PossiblyUpdatingExpression';
+import { errXPTY0004 } from './xquery/XQueryErrors';
+
+function getFirstPrimitiveType(values: Value[]): string | null {
+	const firstActualValue = values.find((value) => !!value);
+
+	if (!firstActualValue) {
+		return null;
+	}
+
+	return getPrimitiveTypeName(firstActualValue.type);
+}
+
+function tryCastToSingleType(values: Value[]): Value[] {
+	const areAllOfStringOrAnyUriType = values.every((value) => {
+		if (!value) {
+			return true;
+		}
+
+		const primitiveType = getPrimitiveTypeName(value.type);
+		return primitiveType === 'xs:string' || primitiveType === 'xs:anyURI';
+	});
+
+	if (areAllOfStringOrAnyUriType) {
+		return values.map((value) => (value ? castToType(value, 'xs:string') : value));
+	}
+
+	const areAllOfDecimalOrFloatType = values.every((value) => {
+		if (!value) {
+			return true;
```

Comment:
Same for this, when does this happen? How can there be null/undefined in the values array?